pkg/plugins/runtime/k8s/controllers/gatewayapi/attachment/attachment.go

```diff
@@ -0,0 +1,181 @@
+package attachment
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	kube_apierrs "k8s.io/apimachinery/pkg/api/errors"
+	kube_apimeta "k8s.io/apimachinery/pkg/api/meta"
+	kube_meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kube_labels "k8s.io/apimachinery/pkg/labels"
+	kube_types "k8s.io/apimachinery/pkg/types"
+	kube_client "sigs.k8s.io/controller-runtime/pkg/client"
+	gatewayapi "sigs.k8s.io/gateway-api/apis/v1alpha2"
+
+	"github.com/kumahq/kuma/pkg/plugins/runtime/k8s/controllers/gatewayapi/common"
+)
+
+type Attachment int
+
+const (
+	// Unknown means we shouldn't report on whether it's attached or not
+	Unknown Attachment = iota
+	// NotPermitted means the route isn't allowed to attach to the ref
+	NotPermitted
+	// Invalid means the route points to a nonexistent parent
+	// or is otherwise invalid
+	Invalid
+	// Attached means we can successfully attached
+	Attached
+)
+
+func listenerReady(
+	statuses []gatewayapi.ListenerStatus,
+	name gatewayapi.SectionName,
+) bool {
+	for _, status := range statuses {
+		if status.Name != name {
+			continue
+		}
+
+		return kube_apimeta.IsStatusConditionTrue(status.Conditions, string(gatewayapi.ListenerConditionReady))
+	}
+
+	return false
+}
+
+// findRouteListenerAttachment reports whether this ref is allowed to attach to
+// the Gateway or specific Listener.
+func findRouteListenerAttachment(
+	gateway *gatewayapi.Gateway,
+	routeNs kube_client.Object,
+	refSectionName *gatewayapi.SectionName,
+) (Attachment, error) {
+	// Build a map of whether attaching to each listener is possible
+	listeners := map[gatewayapi.SectionName]Attachment{}
+
+	for _, l := range gateway.Spec.Listeners {
+		ns := l.AllowedRoutes.Namespaces
+
+		// TODO
+		// The gateway controller ensures Kinds is either empty or contains
+		// HTTPRoute, so we can ignore it here.
+
+		switch *ns.From {
+		case gatewayapi.NamespacesFromSelector:
+			// TODO, the gateway controller/webhook should verify this
+			selector, err := kube_meta.LabelSelectorAsSelector(ns.Selector)
+			if err != nil {
+				return Unknown, errors.Wrap(err, "internal error: couldn't convert to selector")
+			}
+
+			if !selector.Matches(kube_labels.Set(routeNs.GetLabels())) {
+				listeners[l.Name] = NotPermitted
+				continue
+			}
+		case gatewayapi.NamespacesFromSame:
+			if gateway.Namespace != routeNs.GetName() {
+				listeners[l.Name] = NotPermitted
+				continue
+			}
+		case gatewayapi.NamespacesFromAll:
+		}
+
+		if listenerReady(gateway.Status.Listeners, l.Name) {
```

Comment:
It's not clear to me in the spec what should be the ParentStatus if the parent Listener doesn't have a Ready status yet.

The idea was to be conservative and not set the ParentStatus unless the parent is ready (or the attachment isn't permitted). For the "whole Gateway" case below, the attachment would probably also need to only be set to `NotPermitted` when _every_ listener is `NotPermitted`. Otherwise `Unknown`.

Reply:
You're right, I'm removing that check